docs/docs/Best_Practices/Message_Limits.md

```diff
@@ -0,0 +1,27 @@
+# Message Limits
+
+There are limits on how much data can be sent back and forth between the plugin and engine at a time. There are five scenarios where this comes into play:
+1. Inputs sent from plugin to engine, as arguments to a [Plugin Operation](/References/Plugin_Operations.md). For example, the schema-defined `Repository` object that is provided as input to plugin operations.
+2. Outputs sent back from the plugin to the engine, as the return values from plugin operations.
+3. Exception messages and call stacks thrown by plugin code. For example, the `message` field within [User Visible Errors](/Best_Practices/User_Visible_Errors.md).
+4. Inputs sent from plugin to engine, as arguments to a [Platform library](/References/Platform_Libraries.md) function. For example, the `message` field that is passed to `logger.debug`. 
+5. Outputs sent back from the plugin to the engine, as the return values from Platform Library functions. For example, the `stdout` resulting from a call to `libs.run_bash`. 
+
+For case 1 and 2, the total size of data must be less than 4 mebibytes (4 MiB).
+
+For case 3, the total size of data must be less than 128 kibibytes (128 KiB).
+
+For case 4 and 5, the total size of data must be less than 32 mebibytes (32 MiB).
+
+The actual size of this information at runtime is dependent on how the Python interpreter chooses to represent the information, so it's not always possible to know ahead of time what the exact size will be.
+
+Here are some examples of where problems may occur:
+1. Using `libs.run_bash` to print the entire contents of a large file to stdout.
+2. Using a single `logger` command with many pages of output.
+3. Throwing an exception with a large message or stack trace.
```

Comment:
lets also add something like to account for the large plugin object case?

> 4. Large amount of metadata in a plugin defined schema like "Repository" or "Virtual Source"

Reply:
Just to be a pedant, this should probably be something like "Large amount of data in an object defined by a plugin schema, such as "Repository" or "Virtual Source".   (It's the object that contains lots of data, not the schema)